File: java/com/dragon/symbols/Type.java
package com.dragon.symbols;

import com.dragon.lexer.Tag;
import com.dragon.lexer.Word;

public class Type extends Word {
  public static final Type Int = new Type("int", Tag.BASIC, 4);
  public static final Type Float = new Type("float", Tag.BASIC, 8);
  public static final Type Char = new Type("char", Tag.BASIC, 1);
  public static final Type Bool = new Type("bool", Tag.BASIC, 1);

  public static boolean isNumeric(Type p) {
    return p == Type.Char || p == Type.Int || p == Type.Float;
  }

  public static Type max(Type left, Type right) {
    if (!isNumeric(left) || !isNumeric(right)) {
      return null;
    }
    if (left == Type.Float || right == Type.Float) {
      return Type.Float;
    }
    if (left == Type.Int || right == Type.Int) {
      return Type.Int;
    }

    return Type.Char;
  }

  private int width = 0;

  public Type(String s, int tag, int w) {
    super(s, tag);
    width = w;
  }

  public int getWidth() {
    return width;
  }
}

File: cpp/symbols/type.cpp
#include <symbols/type.hpp>

#include <lexer/token.hpp>

namespace symbols {
std::shared_ptr<Type> Type::integer = Type::create("int", Token::kBasic, 4);
std::shared_ptr<Type> Type::real = Type::create("float", Token::kBasic, 8);
std::shared_ptr<Type> Type::character = Type::create("char", Token::kBasic, 1);
std::shared_ptr<Type> Type::boolean = Type::create("bool", Token::kBasic, 1);

Type *Type::max(Type *type_left, Type *type_right) {
    if (!type_left)
        return type_right;
    else if (!type_right)
        return type_left;
    else if (!type_left->is_numeric() || !type_right->is_numeric())
        return nullptr;
    else if (*type_left == *real || *type_right == *real)
        return real.get();
    else if (*type_left == *integer || *type_right == *integer)
        return integer.get();
    else
        return character.get();
}

bool Type::operator==(const Type &other) const {
    if (this == &other)
        return true;
    return width_ == other.width_ && this->lexeme() == other.lexeme() && this->tag() == other.tag();
}

bool Type::is_numeric() const {
    return (*real == *this) || (*integer == *this) || (*this == *character);
}
} // namespace symbols
